### ISSUE TITLE
choose better default CRAN repository for out-of-date releases

### ISSUE THREAD

**Issue by mtmorgan** [Bioconductor/BiocManager#71]
By default BiocManager always uses the 'current' CRAN repository, but this is incorrect for out-of-date Bioconductor releases, https://github.com/Bioconductor/BiocManager/issues/69#issuecomment-629705944. One possibility is to amend https://bioconductor.org/config.yaml with best-guess repositories for each release, and use these to populate `BiocManager::repositories()[["CRAN"]]`.

**Comment by slopp** [comment]
Hi @mtmorgan, we'd love to help with this request, providing an option for users to opt into using a compatible snapshot from https://packagemanager.rstudio.com. Would you be open to a PR?  

We were thinking it would be helpful for users to be able to specify a CRAN-checkpoint base (whether packagemanager.rstudio.com or an on-prem RSPM instance) and then we could append the appropriate checkpoint CRAN repo based on the version of R and Bioconductor. 

For air-gapped users, it would also be helpful if there were an option to specify where the config.yml file lives (in cases where it could be mirrored and hosted offline).

Happy to talk through this as well if you'd like.

cc @jonyoder

**Comment by mtmorgan** [comment]
@slopp @jonyoder @LiNk-NY would be great to have a short call to talk through this, in the 2-4 window (US Eastern) today or tomorrow; I can figure out everyone's email but @jonjoder to send an invite...

**Comment by jonyoder** [comment]
@mtmorgan I'm at jonathan.yoder@rstudio.com

**Comment by slopp** [comment]
A potential proposal 

```
options(freeze_cran = rspm | mran | rspm_mirror | false)
```

The values could mean something like:

- `rspm`: adds a CRAN repo that points at the correct `https://packagemanager.rstudio.com` checkpoint based on the release. We could attempt to point at a Linux OS binary URL if determinable
- `mran`: adds a CRAN repo that points at the correct MRAN checkpoint based on the release
- `rspm_mirror`: the same as the rspm option, but inherits the bioc_mirror base instead of using packagemanager.rstudio.com as the base
- `false`: this would be the default and would retain the current behavior of just appending the current repo option for CRAN

These options would be supported by the config.yml file that maps Bioc releases/dates to immutable URLs for the different services. 

We could go further, if we wanted, and recommend a service (mran or rspm) based on the OS.

In addition, we'd like to see if it'd be possible to host the `config.yml` file off of the bioc_mirror base to better support totally offline mirrors.

Note that this behavior would require changes to  RSPM, which doesn't support all of this yet today. And we'd want to determine how to support and maintain the config mapping.

**Comment by kevinushey** [comment]
Just to repeat all the work items thus far (to make sure I understand it correctly):

1. The file at https://bioconductor.org/config.yaml would be amended, to include information on the appropriate RSPM / MRAN repository URLs that could be used for a particular Bioconductor release. The selection of these URLs would be done in collaboration with the Bioconductor / MRAN / RStudio teams, and RSPM would take the time to curate stable repository URLs to be used in these cases.

2. Users could configure which of the aforementioned URLs are used as the "CRAN" repository URL. This would be done through an R option; ostensibly called `bioc_freeze_cran`.

`BiocManager` already allows one to customize the Bioconductor URL via the `BioC_mirror` option, so I think there's no changes required on that front.

The changes would likely occur here:

https://github.com/Bioconductor/BiocManager/blob/f0b76f8979f513ac0cfcf4b1485dacba0c6e5dcb/R/repositories.R#L44-L52

In particular, rather than using `getOption("repos")` as the default "base" repository, we'd instead use one of the MRAN / RSPM repository URLs as appropriate.

**Comment by slopp** [comment]
@kevinushey those requirements look correct to me! 

The one additional note I'd make, `Bioc_mirror` changes the base for Bioconductor repositories, it would be awesome if we could also respect a RSPM-base for CRAN somehow. For example, if the user wants to pull CRAN and Bioconductor packages from RSPM, we'd want to be careful not to over-write the CRAN url with `https://packagemanager.rstudio.com/cran/...path..to..correct...snapshot`, but with `their/rspm/server/repo/...path..to....`

**Comment by kevinushey** [comment]
That makes sense. We could get the "base" path to RSPM with something like:

```
getOption("RSPM_mirror", default = "https://packagemanager.rstudio.com")
```

and use that to allow users to point towards a separate RSPM instance if so required.

Allowing users to choose a "default" RSPM path in an R session seems like something worth considering beyond just `BiocManager` so we should probably discuss if this is the "right" way to do this globally. (ie: could have implications for how `renv` uses RSPM as well, and other packages)

**Comment by LiNk-NY** [comment]
This has been implemented here ab16fc2f19bc8a50084e369b85dd3f72682bb9a9 though the user has to opt-in on the feature 7600bff460a3b375de391ab7e7823601a7c803f6. Feel free to open a new issue if warranted.
